src/ZPublisher/cookie.py

```diff
@@ -0,0 +1,337 @@
+##############################################################################
+#
+# Copyright (c) 2022 Zope Foundation and Contributors.
+#
+# This software is subject to the provisions of the Zope Public License,
+# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
+# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
+# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
+# FOR A PARTICULAR PURPOSE.
+#
+##############################################################################
+"""HTTP cookie support.
+
+The implementation is based on
+`RFC6265 <https://www.rfc-editor.org/rfc/rfc6265.html>`_.
+
+The module introduces 2 features:
+
+ 1. default ``ICookieParamPolicy`` and ``ICookieValuePolicy``
+ 2. a cookie parameter registry
+
+and exports the functions ``getCookieParamPolicy``, ``getCookieValuePolicy``
+``registerCookieParameter``,
+``normalizeCookieParameterName`` and ``convertCookieParameter``.
+"""
+import datetime
+from itertools import chain
+from re import compile
+from time import time
+
+from six.moves.urllib.parse import quote
+from six.moves.urllib.parse import unquote
+
+from DateTime import DateTime
+from zope.component import getGlobalSiteManager
+from zope.component import queryUtility
+from zope.interface import implementer
+
+from .interfaces import ICookieParamPolicy
+from .interfaces import ICookieValuePolicy
+
+
+@implementer(ICookieParamPolicy)
+class DefaultCookieParamPolicy(object):
+    """Default ``ICookieParamPolicy``.
+
+    It adds ``Expires`` when it sees ``Max-Age`` (for compatibility
+    with HTTP/1.0) and urlquotes cookie values.
+    """
+
+    @staticmethod
+    def parameters(name, attrs):
+        """adds ``Expires`` if not present and ``Max-Age`` is set."""
+        for pn, pv in attrs.items():
+            if pn != "value":
+                yield (pn, pv)
+        if "Expires" not in attrs and attrs.get("Max-Age") is not None:
+            max_age = int(attrs["Max-Age"])
+            expires = "Wed, 31 Dec 1997 23:59:59 GMT" if max_age <= 0 \
+                      else time() + max_age
+            yield convertCookieParameter("Expires", expires)
+
+    @staticmethod
+    def check_consistency(name, attrs):
+        return  # does not enforce any consistency rules
```

Comment:
A little higher up a very old comment in the `_cookie_list` response method says "should really check size of cookies here". If the RFC mentions a maximum size wouldn't it make sense to check it here?

Reply:
RFC 6265 does not specify a cookie size limit. In section 6.1, it recommends that a browser should support cookies of at least up to 4096 bytes size. Not sure whether this should cause us to reject larger cookies.